File: RogueTraderSystemGenerator/Nodes/DerelictStationNode.cs
using System;
using System.Globalization;
using System.Runtime.Serialization;
using System.Windows.Documents;

namespace RogueTraderSystemGenerator.Nodes
{
    [DataContract]
    class DerelictStationNode : NodeBase
    {
        [DataMember]
        private const int Armour = 10;
        [DataMember]
        private string _stationOrigin;
        [DataMember]
        private int _hullIntegrity;

        public override void ResetVariables()
        {
            base.ResetVariables();
            _stationOrigin = "";
            _hullIntegrity = 0;
        }

        public DerelictStationNode(SystemCreationRules systemCreationRules)
        {
            _nodeName = "Derelict Station";
            _systemCreationRules = systemCreationRules;
        }

        public override void GenerateFlowDocument()
        {
            _flowDocument = new FlowDocument();
            DocBuilder.AddHeader(ref _flowDocument, NodeName, 3);
            if (!String.IsNullOrEmpty(Description))
                DocBuilder.AddContentLine(ref _flowDocument, "Description", new DocContentItem(Description));
            DocBuilder.AddContentLine(ref _flowDocument, "Station Type", new DocContentItem(_stationOrigin, 15, "Table 1-4: Derelict Station Origins"));
            DocBuilder.AddContentLine(ref _flowDocument, "Armour", new DocContentItem(Armour.ToString(CultureInfo.InvariantCulture), 15, "Derelict Station"));
            DocBuilder.AddContentLine(ref _flowDocument, "Hull Integrity", new DocContentItem(_hullIntegrity.ToString(CultureInfo.InvariantCulture), 15, "Derelict Station"));
            if (_resourceArcheotechCache > 0)
                DocBuilder.AddContentLine(ref _flowDocument, "Archeotech Resources", new DocContentItem(ResourceArcheotechCacheText));
            if (_resourceXenosRuins > 0)
                DocBuilder.AddContentLine(ref _flowDocument, "Xenotech Resources", new DocContentItem(ResourceXenosRuinsText));
            if (_resourceArcheotechCache == 0 && _resourceXenosRuins == 0)
                DocBuilder.AddContentLine(ref _flowDocument, "Resources", new DocContentItem("None"));

            if (_inhabitants != Species.None)
            {
                DocBuilder.AddContentLine(ref _flowDocument, "Inhabitants", new DocContentItem(GetSpeciesText(_inhabitants)));
                DocBuilder.AddContentLine(ref _flowDocument, "Inhabitant Development", _inhabitantDevelopment);
            }
        }

        public override void Generate()
        {
            bool xenos = false; // If false, generate Archeotech. Else generate Xenotech.
            int randValue = Globals.RollD100();
            // Override random value in certain cases to enable the dominant species for the system
            if(_systemCreationRules.DominantRuinedSpecies != XenosRuinsSpecies.Undefined && Globals.RollD10() <= 5)
            {
                switch (_systemCreationRules.DominantRuinedSpecies)
                {
                    case XenosRuinsSpecies.Undefined:
                    case XenosRuinsSpecies.YuVath:
                    case XenosRuinsSpecies.Kroot:
                        break;
                    case XenosRuinsSpecies.UndiscoveredSpecies:
                        randValue = 77 + Globals.Rand.Next(22);
                        break;
                    case XenosRuinsSpecies.Eldar:
                        randValue = 11 + Globals.Rand.Next(14);
                        break;
                    case XenosRuinsSpecies.Egarian:
                        randValue = 5;
                        break;
                    case XenosRuinsSpecies.Ork:
                        randValue = 39;
                        break;
                    default:
                        throw new ArgumentOutOfRangeException();
                }
            }
            if (randValue <= 10)
            {
                _stationOrigin = "Egarian Void-maze";
                xenos = true;
                if(_systemCreationRules.DominantRuinedSpecies == XenosRuinsSpecies.Undefined && Globals.RollD10() <= 6)
                    _systemCreationRules.DominantRuinedSpecies = XenosRuinsSpecies.Egarian;
            }
            else if (randValue <= 20)
            {
                _stationOrigin = "Eldar Orrery";
                xenos = true;
                if (_systemCreationRules.DominantRuinedSpecies == XenosRuinsSpecies.Undefined && Globals.RollD10() <= 6)
                    _systemCreationRules.DominantRuinedSpecies = XenosRuinsSpecies.Eldar;
            }
            else if (randValue <= 25)
            {
                _stationOrigin = "Eldar Gate";
                xenos = true;
                if (_systemCreationRules.DominantRuinedSpecies == XenosRuinsSpecies.Undefined && Globals.RollD10() <= 6)
                    _systemCreationRules.DominantRuinedSpecies = XenosRuinsSpecies.Eldar;
            }
            else if (randValue <= 40)
            {
                _stationOrigin = "Ork Rok";
                xenos = true;
                if (_systemCreationRules.DominantRuinedSpecies == XenosRuinsSpecies.Undefined && Globals.RollD10() <= 6)
                    _systemCreationRules.DominantRuinedSpecies = XenosRuinsSpecies.Ork;
            }
            else if (randValue <= 50)
            {
                _stationOrigin = "STC Defence Station";
            }
            else if (randValue <= 65)
            {
                _stationOrigin = "STC Monitor Station";
            }
            else if (randValue <= 76)
            {
                _stationOrigin = "Stryxis Collection";
                xenos = true;
            }
            else if (randValue <= 85)
            {
                _stationOrigin = "Xenos Defence Station";
                xenos = true;
                if (_systemCreationRules.DominantRuinedSpecies == XenosRuinsSpecies.Undefined && Globals.RollD10() <= 6)
                    _systemCreationRules.DominantRuinedSpecies = XenosRuinsSpecies.UndiscoveredSpecies;
            }
            else
            {
                _stationOrigin = "Xenos Monitor Station";
                xenos = true;
                if (_systemCreationRules.DominantRuinedSpecies == XenosRuinsSpecies.Undefined && Globals.RollD10() <= 6)
                    _systemCreationRules.DominantRuinedSpecies = XenosRuinsSpecies.UndiscoveredSpecies;
            }

            int numResources = Globals.RollD5() - 1;
            for (int i = 0; i < numResources; i++)
            {
                if (xenos)
                {
                    _resourceXenosRuins += Globals.RollD100();
                    if (_systemCreationRules.RuinedEmpireIncreasedAbundanceXenosRuins)
                        _resourceXenosRuins += (Globals.RollD10() + 5);
                }
                else
                {
                    _resourceArcheotechCache += Globals.RollD100();
                    if (_systemCreationRules.RuinedEmpireIncreasedAbundanceArcheotechCaches)
                        _resourceArcheotechCache += (Globals.RollD10() + 5);
                }
            }
            _hullIntegrity = Globals.RollD10(4);
        }
    }
}

File: electron-app/js/nodes/derelictStationNode.js
// DerelictStationNode.js - Parity rewrite with C# DerelictStationNode
// Implements Table 1-4: Derelict Station Origins, hull integrity, and Ruined Empire resource logic.
class DerelictStationNode extends NodeBase {
    constructor(id = null) {
        super(NodeTypes.DerelictStation, id);
        this.nodeName = 'Derelict Station';
        this.fontWeight = 'bold';
        this.stationOrigin = '';
        this.hullIntegrity = 0; // 4d10 total (parity)
        this.armor = 10; // Fixed armour value
        this.systemCreationRules = null; // injected from parent SystemNode via ZoneNode helper
        // Resource modeling (arrays of { type, abundance }) to align with planet parity & Ruined Empire injections
        this.xenosRuins = [];
        this.archeotechCaches = [];
        // Inhabitants fields inherited (inhabitants / inhabitantDevelopment) used by Starfarers distribution
    }

    generate() {
        super.generate();
        this.pageReference = createPageReference(15, 'Table 1-4: Derelict Station Origins');
        // Reset per-regeneration state
        this.xenosRuins = [];
        this.archeotechCaches = [];
        this.hullIntegrity = 0;
        this._generateOriginAndResources();
        this._generateHullIntegrity();
        this.updateDescription();
    }

    _generateOriginAndResources() {
        // Bias towards existing dominant ruined species (if already chosen) 50% chance (RollD10 <=5)
        let roll = RollD100();
        const rules = this.systemCreationRules;
        const dom = rules?.dominantRuinedSpecies;
        if (dom && RollD10() <= 5) {
            switch (dom) {
                case 'Undiscovered': // Map to Undiscovered Species (generic xenos) -> pick range 77-98 (covers defence/monitor majority)
                    roll = 77 + RandBetween(0, 21); // 77-98 inclusive
                    break;
                case 'Eldar':
                    roll = 11 + RandBetween(0, 13); // 11-24
                    break;
                case 'Egarian':
                    roll = 5; // within <=10 bracket
                    break;
                case 'Ork':
                    roll = 39; // within 26-40 bracket
                    break;
                case 'Yu' + "'Vath": // not chosen here normally, leave default
                case 'Kroot':
                default:
                    break; // no special override
            }
        }

        let isXenos = false; // Determines whether resource packets are Xenos Ruins vs Archeotech Caches
        // Origin table (roll ranges) parity with C#
        if (roll <= 10) {
            this.stationOrigin = 'Egarian Void-maze'; isXenos = true; this._maybeSetDominant('Egarian');
        } else if (roll <= 20) {
            this.stationOrigin = 'Eldar Orrery'; isXenos = true; this._maybeSetDominant('Eldar');
        } else if (roll <= 25) {
            this.stationOrigin = 'Eldar Gate'; isXenos = true; this._maybeSetDominant('Eldar');
        } else if (roll <= 40) {
            this.stationOrigin = 'Ork Rok'; isXenos = true; this._maybeSetDominant('Ork');
        } else if (roll <= 50) {
            this.stationOrigin = 'STC Defence Station';
        } else if (roll <= 65) {
            this.stationOrigin = 'STC Monitor Station';
        } else if (roll <= 76) {
            this.stationOrigin = 'Stryxis Collection'; isXenos = true; // Does not set dominant species in C#
        } else if (roll <= 85) {
            this.stationOrigin = 'Xenos Defence Station'; isXenos = true; this._maybeSetDominant('Undiscovered');
        } else {
            this.stationOrigin = 'Xenos Monitor Station'; isXenos = true; this._maybeSetDominant('Undiscovered');
        }

        // Resource packets: (RollD5 -1) iterations, each adds a d100 abundance (+ optional Ruined Empire bonus d10+5)
        const num = RollD5() - 1;
        for (let i = 0; i < num; i++) {
            const abundanceBase = RollD100();
            let abundance = abundanceBase;
            if (isXenos && rules?.ruinedEmpireIncreasedAbundanceXenosRuins) abundance += RollD10() + 5;
            if (!isXenos && rules?.ruinedEmpireIncreasedAbundanceArcheotechCaches) abundance += RollD10() + 5;
            if (isXenos) {
                if (!this.xenosRuins) this.xenosRuins = [];
                const type = this.generateXenosRuins();
                this.xenosRuins.push({ type, abundance });
            } else {
                if (!this.archeotechCaches) this.archeotechCaches = [];
                const type = this.generateArcheotechCache();
                this.archeotechCaches.push({ type, abundance });
            }
        }
    }

    _maybeSetDominant(species) {
        const rules = this.systemCreationRules;
        if (!rules) return;
        if (!rules.dominantRuinedSpecies || rules.dominantRuinedSpecies === 'Undefined') {
            if (RollD10() <= 6) rules.dominantRuinedSpecies = species; // 60% chance parity (RollD10 <=6)
        }
    }

    _generateHullIntegrity() {
        this.hullIntegrity = RollD10(4); // 4d10
    }

    generateArcheotechCache() {
        const types = [
            'Ancient Data Repository',
            'Technological Ruins',
            'Archeotech Device Cache',
            'Pre-Age of Strife Facility',
            'Dark Age Technology'
        ];
        return ChooseFrom(types);
    }

    generateXenosRuins() {
        // Parity: Same species weighting / dominance adoption as WPF NodeBase.GenerateXenosRuins.
        const rules = this.systemCreationRules;
        const adoptDominant = rules && rules.dominantRuinedSpecies && rules.dominantRuinedSpecies !== 'Undefined' && RollD10() <= 6;
        let speciesKey;
        if (adoptDominant) {
            speciesKey = rules.dominantRuinedSpecies;
        } else {
            const r = RollD10();
            if (r <= 4) speciesKey = 'Undiscovered';
            else if (r <= 6) speciesKey = 'Eldar';
            else if (r === 7) speciesKey = 'Egarian';
            else if (r === 8) speciesKey = "Yu'Vath";
            else if (r === 9) speciesKey = 'Ork';
            else speciesKey = 'Kroot';
            if (rules && (!rules.dominantRuinedSpecies || rules.dominantRuinedSpecies === 'Undefined')) {
                if (RollD10() <= 7) rules.dominantRuinedSpecies = speciesKey;
            }
        }
        switch (speciesKey) {
            case 'Eldar': return 'Eldar Ruins';
            case 'Egarian': return 'Egarian Remains';
            case "Yu'Vath": return "Yu'Vath Structures";
            case 'Ork': return 'Ork Settlements';
            case 'Kroot': return 'Kroot Encampments';
            case 'Undiscovered':
            default: return 'Undiscovered Species';
        }
    }

    updateDescription() {
        let desc = `<h3>Derelict Station</h3>`;
        const showPages = window.APP_STATE?.settings?.showPageNumbers;
        const originRef = showPages ? ` <span class="page-reference">${createPageReference(15, 'Table 1-4: Derelict Station Origins')}</span>` : '';
        const statRef = showPages ? ` <span class="page-reference">${createPageReference(15, 'Derelict Station')}</span>` : '';
        desc += `<p><strong>Station Type:</strong> ${this.stationOrigin}${originRef}</p>`;
        desc += `<p><strong>Armour:</strong> ${this.armor}${statRef}</p>`;
        desc += `<p><strong>Hull Integrity:</strong> ${this.hullIntegrity}${statRef}</p>`;

        // Resources
        const archeo = (this.archeotechCaches || []).filter(a=>a.abundance>0);
        const xenos = (this.xenosRuins || []).filter(r=>r.abundance>0);
        if (archeo.length>0) {
            desc += '<h4>Archeotech Resources</h4><ul>' + archeo.map(a=>`<li>${a.type} (Abundance ${a.abundance})</li>`).join('') + '</ul>';
        }
        if (xenos.length>0) {
            desc += '<h4>Xenotech Resources</h4><ul>' + xenos.map(r=>`<li>${r.type} (Abundance ${r.abundance})</li>`).join('') + '</ul>';
        }
        if (archeo.length===0 && xenos.length===0) {
            desc += '<p><strong>Resources:</strong> None</p>';
        }

        if (this.inhabitants && this.inhabitants !== 'None') {
            desc += `<p><strong>Inhabitants:</strong> ${this.inhabitants}</p>`;
            if (this.inhabitantDevelopment) desc += `<p><strong>Inhabitant Development:</strong> ${this.inhabitantDevelopment}</p>`;
        }

        this.description = desc;
    }

    toJSON() {
        const json = super.toJSON();
        Object.assign(json, {
            stationOrigin: this.stationOrigin,
            hullIntegrity: this.hullIntegrity,
            armor: this.armor,
            xenosRuins: this.xenosRuins,
            archeotechCaches: this.archeotechCaches,
            inhabitants: this.inhabitants,
            inhabitantDevelopment: this.inhabitantDevelopment
        });
        return json;
    }

    static fromJSON(data) {
        const node = new DerelictStationNode(data.id);
        
        // Restore base properties
        Object.assign(node, {
            nodeName: data.nodeName || 'Derelict Station',
            description: data.description || '',
            customDescription: data.customDescription || '',
            pageReference: data.pageReference || '',
            isGenerated: data.isGenerated || false,
            fontWeight: data.fontWeight || 'normal',
            fontStyle: data.fontStyle || 'normal',
            fontForeground: data.fontForeground || '#000000'
        });
        
        // Restore derelict station-specific properties
        node.stationOrigin = data.stationOrigin || '';
        node.hullIntegrity = data.hullIntegrity || 0;
        node.armor = data.armor || 10;
        node.xenosRuins = data.xenosRuins || [];
        node.archeotechCaches = data.archeotechCaches || [];
        node.inhabitants = data.inhabitants || 'None';
        node.inhabitantDevelopment = data.inhabitantDevelopment || '';
        
        return node;
    }
}

window.DerelictStationNode = DerelictStationNode;